Client/Frontend/Browser/HomePanel/NTPDataSource.swift

```diff
@@ -0,0 +1,242 @@
+    
+    private lazy var standardBackgrounds: [NTPBackground] = {
+        let backgroundFilePath = "ntp-data"
+        guard let backgroundData = self.loadData(file: backgroundFilePath) else { return [] }
+        
+        do {
+            return try JSONDecoder().decode([NTPBackground].self, from: backgroundData)
+        } catch {
+            return []
+        }
+    }()
+    
+    init() {
+        downloader.delegate = self
+        
+        Preferences.NewTabPage.backgroundSponsoredImages.observe(from: self)
+        Preferences.NewTabPage.selectedCustomTheme.observe(from: self)
+    }
+    
+    func startFetching() {
+        if downloader.delegate != nil {
+            downloader.notifyObservers(for: downloader.currentResourceType)
+        }
+    }
+    
+    func fetchSpecificResource(_ type: NTPDownloader.ResourceType) {
+        if downloader.delegate != nil {
+            downloader.notifyObservers(for: type)
+        }
+    }
+    
+    // This is used to prevent the same handful of backgrounds from being shown.
+    //  It will track the last N pictures that have been shown and prevent them from being shown
+    //  until they are 'old' and dropped from this array.
+    // Currently only supports normal backgrounds, as sponsored images are not supposed to be duplicate.
+    // This can 'easily' be adjusted to support both sets by switching to String, and using filePath to identify uniqueness.
+    private var lastBackgroundChoices = [Int]()
+    
+    enum BackgroundType {
+        case regular
+        case withBrandLogo(_ logo: NTPLogo?)
+        case withQRCode(_ code: String)
+    }
+    
+    private enum ImageRotationStrategy {
+        /// Special strategy for sponsored images, uses in-memory property to keep track of which image to show.
+        case sponsoredRotation
+        /// Uses random images, keeps in-memory list of recently viewed images to avoid showing it too often.
+        case randomOrderAvoidDuplicates
+    }
+    
+    func newBackground() -> (NTPBackground, BackgroundType)? {
+        if !Preferences.NewTabPage.backgroundImages.value { return nil }
+        
+        // Identifying the background array to use
+        let (backgroundSet, backgroundType, strategy) = {
+            () -> ([NTPBackground], BackgroundType, ImageRotationStrategy) in
+            
+            if let theme = customTheme, let refCode = theme.refCode,
+                Preferences.NewTabPage.selectedCustomTheme.value != nil {
+                return (theme.wallpapers, .withQRCode(refCode), .randomOrderAvoidDuplicates)
+            }
+            
+            if let sponsor = sponsor {
+                let attemptSponsored = Preferences.NewTabPage.backgroundSponsoredImages.value
+                    && backgroundRotationCounter == NTPDataSource.sponsorshipShowValue
+                    && !PrivateBrowsingManager.shared.isPrivateBrowsing
+                
+                return attemptSponsored ?
```

Comment:
Since the back condition of this is identical to the fallback return, just remove it, since it will drop down and just perform the return [here](https://github.com/brave/brave-ios/pull/2408/files#diff-4b4b5010b1b99252e9b24c6869532871R125)

```suggestion
                if attemptSponsor {
                    return (sponsor.wallpapers, .withBrandLogo(sponsor.logo), .sponsoredRotation)
                }
```